crates/burn-dataset/src/transform/selection.rs

```diff
@@ -0,0 +1,233 @@
+use crate::Dataset;
+use rand::SeedableRng;
+use rand::prelude::SliceRandom;
+use rand::rngs::StdRng;
+use std::marker::PhantomData;
+
+/// Generates a vector of indices from 0 to size - 1.
+///
+/// # Arguments
+///
+/// * `size` - The size of the dataset.
+///
+/// # Returns
+///
+/// A vector containing indices from 0 to size - 1.
+#[inline(always)]
+pub fn iota(size: usize) -> Vec<usize> {
+    (0..size).collect()
+}
+
+/// Generates a shuffled vector of indices up to a size.
+///
+/// # Arguments
+///
+/// * `size` - The size of the dataset to shuffle.
+///
+/// # Returns
+///
+/// A vector of shuffled indices.
+#[inline(always)]
+pub fn shuffled_indices(size: usize, rng: &mut StdRng) -> Vec<usize> {
+    let mut indices = iota(size);
+    indices.shuffle(rng);
+    indices
+}
+
+/// A dataset that selects a subset of indices from an existing dataset.
+///
+/// Indices may appear multiple times, but they must be within the bounds of the original dataset.
+pub struct SelectionDataset<D, I> {
+    dataset: D,
+    indices: Vec<usize>,
+    input: PhantomData<I>,
+}
+
+impl<D, I> SelectionDataset<D, I>
+where
+    D: Dataset<I>,
+{
+    /// Creates a new selection dataset with the given dataset and indices.
+    ///
+    /// # Arguments
+    ///
+    /// * `dataset` - The original dataset to select from.
+    /// * `indices` - A slice of indices to select from the dataset.
+    ///   These indices must be within the bounds of the dataset.
+    ///
+    /// # Panics
+    ///
+    /// Panics if any index is out of bounds for the dataset.
+    pub fn new(dataset: D, indices: Vec<usize>) -> Self {
+        let size = dataset.len();
+
+        if let Some(idx) = indices.iter().find(|&i| *i >= size) {
+            panic!("Index out of bounds for wrapped dataset size: {idx} >= {size}");
+        }
+
+        Self {
+            dataset,
+            indices,
+            input: PhantomData,
+        }
+    }
+
+    /// Creates a new selection dataset that selects all indices from the dataset.
+    ///
+    /// # Arguments
+    ///
+    /// * `dataset` - The original dataset to select from.
+    ///
+    /// # Returns
+    ///
+    /// A new `SelectionDataset` that selects all indices from the dataset.
+    pub fn new_select_all(dataset: D) -> Self {
+        let size = dataset.len();
+        Self::new(dataset, iota(size))
+    }
```

Comment:
It's kind of bad to use, since we create an indices vector for nothing if we don't perform any transformation afterward. It should be in the doc that some transformation should be done for it to be useful.

Reply:
made this much more explicit.